In [1]:
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:100% !important; }</style>"))
display(HTML("<style>.output_result { max-width:100% !important; }</style>"))

C:\Users\pavan\AppData\Local\Temp\ipykernel_15600\4066392039.py:1: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display
  from IPython.core.display import display, HTML


In [2]:
import pandas as pd
import numpy as np

In [23]:
from sklearn import tree
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
import graphviz
from sklearn.preprocessing import StandardScaler

<font size="5">Problem 1</font> <br>
**[Decision tree] Using the breast cancer dataset, let's encode the diagnosis = M  as 1, and diagnosis = B as 0, and treat this as a binary outcome (i.e., dependent variable).**
1) [5 points] What is the value of the gini index of the full dataset? <br>
2) [10 points] If we decide to use gini index as the error metric, and split on the feature of "texture_mean", what is value of the splitting point? (The correct answer falls in a range).<br>
3) [10 points] If we want to repeat part 2, but using entropy as the error metric, what is the value of the splitting point? (The correct answer falls in a range).

In [25]:
data = pd.read_csv("breast_cancer_data.csv")

In [26]:
data.shape

(569, 32)

In [27]:
data.diagnosis = [1 if i=="M" else 0 for i in data.diagnosis] 

In [28]:
X_data = data.drop(['id','diagnosis'],axis = 1)
y_data = data.diagnosis

In [29]:
X_train,X_test, y_train,y_test = train_test_split(X_data,data.diagnosis, test_size=0.2)

In [30]:
DecisionTree = tree.DecisionTreeClassifier(criterion="gini")

In [31]:
DecisionTree = DecisionTree.fit(X_train,y_train)

In [32]:
y_pred = DecisionTree.predict(X_test)

In [33]:
def gini_index(p: float):
    """Gini index for a given binary class ratio."""
    return 2 * p * (1 - p)

In [34]:
print("1.1) The Gini Index for the entire dataset is : {}".format(gini_index(len(data[data.diagnosis == 0])/len(data)))) #arg for gini function is class_ratio

1.1) The Gini Index for the entire dataset is : 0.4675300607546925


In [ ]:
DecisionTree_1b = tree.DecisionTreeClassifier(criterion="gini",splitter="best")

X_data_1b = np.asarray(X_data.texture_mean).reshape(-1, 1)
DecisionTree_1b = DecisionTree_1b.fit(X_data_1b,y_data)

dot_data = tree.export_graphviz(
    decision_tree=DecisionTree_1b, 
    out_file=None, 
    feature_names=['texture_mean'],#X_train.columns,  
    class_names=["0", "1"],  
    filled=True, 
    rounded=True,  
    special_characters=True,
    max_depth=1,
)  
graph = graphviz.Source(dot_data)  
graph.render("cancer_tree")
graph

In [ ]:
print("1.2) The splitting point for texture_mean with gini index as the error criterion is : 19.47")

In [ ]:
DecisionTree_1c = tree.DecisionTreeClassifier(criterion="entropy",splitter="best")

X_data_1c = np.asarray(X_data.texture_mean).reshape(-1, 1)
DecisionTree_1c = DecisionTree_1c.fit(X_data_1c,y_data)

dot_data = tree.export_graphviz(
    decision_tree=DecisionTree_1c, 
    out_file=None, 
    feature_names=['texture_mean'],#X_train.columns,  
    class_names=["0", "1"],  
    filled=True, 
    rounded=True,  
    special_characters=True,
    max_depth=1,
)  
graph = graphviz.Source(dot_data)  
graph.render("cancer_tree")
graph

In [ ]:
print("1.3) The splitting point for texture_mean with entropy as the error criterion is : 18.635")

# Problem 2

**[Neural Network] Using the breast cancer dataset, let's encode the diagnosis = M  as 1, and diagnosis = B as 0, and treat this as a binary outcome (i.e., dependent ariable). Let's use three features, namely,"radius_extreme", "texture_extreme", "perimeter_extreme", to build a neural network (NN). For this NN, let's have one hidden layer, together with the final layer (output layer). For the hidden layer, let's have 5 neurons, and using ReLU as the activation function. For the final layer, let's use sigmoid function as the activations.**<br>
* 1. [5 points] How many parameters are there in this NN? <br>
* 2. [10 points] Using the matrix notation convetion shown in the class, if we have both the matrices  and  filled with the value of 1, and  and  filled with the value of 0.1.After one forward parapagation, what is the value of the average prediction?<br> Note that if there are n  samples, the average prediction is taken over all n  samples.Please standardize the input (i.e., for each feature, subtract its mean, and divide by its standard deviation) before the forward propagation. <br>
* 3. [5 points] Following part 2 above, if we use logloss as the error metric, what is the value of this error metric? Note here the entropy is also taken as the average over all n  samples.

In [ ]:
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import StratifiedKFold
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

In [ ]:
def create_model():
    model = Sequential()
    model.add(Dense(5, input_shape = (None, 3),activation='relu'))
    model.add(Dense(1, activation='sigmoid'))
    # Compile model
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

In [ ]:
model = create_model()

model.summary()

print("\n2.1) Total number of parameters in the model is : 26")

In [ ]:
#using the same dataset since the same attributes are used

In [ ]:
features = [
    "radius_extreme",
    "texture_extreme",
    "perimeter_extreme",
]

In [ ]:
scaler = StandardScaler()
scaler.fit(data[features].values)
X = scaler.transform(data[features].values)
y = data.diagnosis.values.reshape((-1, 1))

In [ ]:
# forward pass for a simple 2-layer NN, with 3 hidden units
np.random.seed(10)

def sigmoid(x):
    """Calculates sigmoid function."""
    return 1. / (1 + np.exp(-x))
def relu(x):
    """Calculates relu function."""
    return np.maximum(0.0,x)

In [ ]:
# parameters for the first layer
W_1 = np.ones(shape=(5, X.shape[1]))
print(f"Shape of W_1 is {W_1.shape}")

b_1 = np.ones(shape=(5, 1))*0.1
print(f"Shape of b_1 is {b_1.shape}")

# parameters for the second layer
W_2 = np.ones(shape=(1, 5))
print(f"Shape of W_2 is {W_2.shape}")

b_2 = np.ones(shape=(1, 1))*0.1
print(f"Shape of b_1 is {b_2.shape}")

# calculate the forward propagation
Z_1 = X @ W_1.T
print(f"\nShape of Z_1 is {Z_1.shape}")
print("Samples for Z_1:")
print(Z_1[:5])

A_1 = relu(Z_1 + b_1.T) 
print(f"Shape of A_1 is {A_1.shape}")
print("Samples for A_1:")
print(A_1[:5])

Z_2 = A_1 @ W_2.T
print(f"\nShape of Z_2 is {Z_2.shape}")
print("Samples for Z_2:")
print(Z_1[:5])

A_2 = Y_hat = sigmoid(Z_2 + b_2.T)
print(f"Shape of A_2 is {A_2.shape}")
print("Samples for A_2:")
print(A_2[:5])

In [ ]:
avg_prediction = sum((A_2)/A_2.shape[0])[0]
print("2.2) The average prediction is : {}".format(avg_prediction))

In [ ]:
log_loss = -np.mean(np.multiply(y, np.log(Y_hat+1E-16)) + np.multiply(1 - y, np.log(1 - Y_hat+1E-16)))
print("2.3) The log loss error metric is : {}".format(log_loss))

# Answers


<font size="5">Problem 1</font>

1) The Gini Index for the entire dataset is : 0.4675300607546925

2) The splitting point for texture_mean with gini index as the error criterion is : 19.47

3) The splitting point for texture_mean with entropy as the error criterion is : 18.635

<font size="5">Problem 2</font>

1) Total number of parameters in the model is : 26

2) The average prediction after one pass is : 0.7134461335701471

3) The log loss error metric is : 0.6811257843182167

<font size="5">Problem 3</font>

Prove that, in Lloyd's algorithm for K-means clustering, for a given cluster  with 
observations, the averaged pairwise within-cluster squared L2 norm equals to the sum of
squared L2 norm of each point (in the cluster) to its cluster center. That is:

$\frac{1}{n}\sum_{i, j\in{C}}\left|\left|x_{i}-x_{j}\right|\right|_{2}^2 = 2\sum_{i \in{C}}\left|\left|x_{i}-\mu\right|\right|_{2}^2 ,$

$where\ \ \mu = \frac{1}{n}\sum_{i}x_{i}\ \ is\ the\ centroid\ of\ the\ cluster.$


<font size="4">Proof : </font>

$ We\ know\ that \ \ \frac{1}{n}\sum_{i}x_{i}\ \ is\ the\ centroid\ of\ the\ cluster.$

$K-Means:$

$ a)\ C_{1},C_{2},C_{3},.......,C_{k} = arg\ min\left(\sum_{k=1}^k W(C_{k})\right)\Longrightarrow(1)$

$$ \ \ \  where\ W(C_{k})\ measures\ within\ cluster\ variations.$$

$ b)\ C_{1}\ \cup\ C_{2}\ \cup\ C_{3}\ \cup\ .......\cup\ C_{k}\ = \left\{ 1,2,3,.......,n \right\}$

$ c)\ C_{k}\ \cap\ C_{k^\prime}\ = \left\{ 1,2,3,.......,n \right\}$

$$ here\ \ C_{1},C_{2},C_{3},.......,C_{k}\ denote\ sets\ containing\ the\ indices\ of\ the\ observations\ of\ each\ cluster.$$


$W(C_{k}) = \frac{1}{|C_{k}|}\sum_{i,i^\prime \in{C_{k}}}\sum_{j=1}^d\left(x_{ij}-x_{i^\prime j}\right)^2 \Longrightarrow(2)$

$K-means\ is\ based\ on\ minimizing\ the\ pairwise\ distance\ of\ data\ points\ within\ the\ same\ cluster.$

$Formally\ given\ x_{1},.....,x_{n}\in R^d,\ we partition\ these\ points\ into\ k\ clusters\ C_{1},......,C_{k}\ based\ on\ objective.$

$Lloyd's\ algorithm:\ \ $

$Substituting\ (2)\ in\ (1)\ :$

$$ \therefore \ \ arg\ min_{C_{1},..,C_{k}} \left(\sum_{i,i^\prime=1}^k\frac{1}{|C_{k}|}\sum_{i,j \in{C_{k}}}\left|\left|x_{ij}-x_{i^\prime j}\right|\right|^2\right)\Longrightarrow(3)$$

$Rewriting\ in\ simple\ terms\ :$

$$ min_{C_{1},..,C_{k}} \left(\sum_{l=1}^k\frac{1}{|C_{l}|}\sum_{i,j \in{C_{l}}}\left|\left|x_{i}-x_{j}\right|\right|^2\right)\Longrightarrow(4)$$

$Here\ C_{1},C_{2},C_{3},.......,C_{k}\ \Longrightarrow\ l = 1,2,....,k. Hence\ written\ as\ C_{l}.$

$Equation\ (4)\ is\ K-means\ in\ a\ simpler\ way.$

$This\ cost\ function\ is\ a\ weighted\ average\ of\ the\ cluster\ variances\ ,with\ weights\ proportional\ $

$to\ the\ cluster\ size\ in\ terms\ of\ number\ of\ points\ |C_{l}| \approx |C_{k}|$

$Let\ \  \mu = \frac{1}{|C_{l}|} \sum_{i \in C_{l}}^n x_{i}\Longrightarrow(5)$

$Splitting\ (4)\ and\ considering\ the\ LHS:$

$$ \therefore \sum_{i,j \in{C_{l}}}\left|\left|x_{i}-x_{j}\right|\right|^2\ = \sum_{i,j \in{C_{l}}}\left(\left|\left|x_{i}\right|\right|^2 + \left|\left|x_{j}\right|\right|^2 -2<x_{i},x_{j}>\right)\Longrightarrow(6)$$

$Applying\ clusters\ in\ (6):$

$$ \sum_{i,j \in{C_{l}}}\left|\left|x_{i}-x_{j}\right|\right|^2\ = \sum_{i\in{C_{l}}}\left(|C_{l}|\left|\left|x_{i}\right|\right|^2 + \sum_{j\in{C_{l}}}\left|\left|x_{j}\right|\right|^2 -2|C_{l}|<x_{i},\mu>\right) $$

$$ = 2|C_{l}|\sum_{i\in{C_{l}}}\left|\left|x_{i}\right|\right|^2 - 2|C_{l}|^2\left|\left|\mu\right|\right|^2\Longrightarrow(7)$$

$$ \therefore = 2|C_{l}|\left[\sum_{i\in{C_{l}}}\left|\left|x_{i}\right|\right|^2 - |C_{l}|\left|\left|\mu\right|\right|^2\right]\Longrightarrow(8)$$

$Considering\ the\ RHS:$

$$\sum_{i \in{C_{l}}}\left|\left|x_{i}-\mu\right|\right|^2\ = \sum_{i\in{C_{l}}}\left(\left|\left|x_{i}\right|\right|^2 + \left|\left|\mu\right|\right|^2 -2<x_{i},\mu>\right)$$

$$ = \sum_{i\in{C_{l}}}\left(\left|\left|x_{i}\right|\right|^2 + \left|\left|C_{l}\right|\right|\left|\left|\mu\right|\right|^2 -2\left|\left|C_{l}\right|\right|\left|\left|\mu\right|\right|^2\right)$$

$$ \therefore \sum_{i \in{C_{l}}}\left|\left|x_{i}-\mu\right|\right|^2\ = \sum_{i\in{C_{l}}}\left(\left|\left|x_{i}\right|\right|^2 -\left|\left|C_{l}\right|\right|\left|\left|\mu\right|\right|^2\right)\Longrightarrow(9)$$

$Substituting\ (9)\ in\ (8):$

$$ \sum_{i,j \in{C_{l}}}\left|\left|x_{i}-x_{j}\right|\right|^2\ = 2|C_{l}|\left[\sum_{i\in{C_{l}}}\left(\left|\left|x_{i}\right|\right|^2 -\left|\left|C_{l}\right|\right|\left|\left|\mu\right|\right|^2\right)\right] $$

$Rearranging\ the\ equation\ and\ substituting\ C_{l} = C_{k} = n,\ we\ get:$

$$ \frac{1}{n}\sum_{i,j \in{C_{l}}}\left|\left|x_{i}-x_{j}\right|\right|_{2}^2\ = 2\sum_{i\in{C_{l}}}\left|\left|x_{i}-\mu\right|\right|_{2}^2\ $$